File: Sorts/Merge Sort/Merge Sort.cpp
#include <iostream> 
#include <vector>
using namespace std;


void merge(int arr[], int l, int m, int r)
{
	int i, j;
	int n1 = m - l + 1;
	int n2 = r - m;
	int L[1000], R[1000];


	for (i = 0; i < n1; i++)
		L[i] = arr[l + i];


	for (j = 0; j < n2; j++)
		R[j] = arr[m + 1 + j];
	
	
	for (i = 0, j = 0;i < n1 && j < n2;)
		arr[l++] = L[i] <= R[j] ? L[i++]: R[j++];
	
	
	while (i < n1)
		arr[l++] = L[i++];
	
	
	while (j < n2)
		arr[l++] = R[j++];
}

void mergeSort(int arr[], int l, int r)
{
	if (l < r)
	{
		int m = l + (r - l) / 2;
		mergeSort(arr, l, m);
		mergeSort(arr, m + 1, r);
		merge(arr, l, m, r);
	}
}


int main(void)
{
	///
	// Exemple of usage
	///
	int *arr;
	int size;
	cout << "n = ";
	cin >> size;

	arr = new int[size];


	///
	// Read array
	///
	for (int i = 0; i < size; i++) {
		cout << "arr[" << i << "] = ";
		cin >> arr[i];
	}
	///
	// Calling a sort function
	///
	mergeSort(arr, 0, size - 1);


	///
	// Output result
	///
	for (int i = 0; i < size; i++) {
		cout << arr[i] << " ";
	}
	cout << endl;
	delete[] arr;
	return 0;
	///
	//  End of exemple of usage
	///
}

File: Sorts/Merge Sort/Merge Sort.py
###
## Start sort function
###

def mergeSort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]
        mergeSort(L)
        mergeSort(R)
        i = j = k = 0
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1
        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

###
## End of sort function
###



###
## Exemple of usage
###

arr = []
size = int(input("n = "))

###
## Read array
###

i = 0
while i < size:
    arr.append(int(input("arr[" + str(i) + "] = ")))
    i += 1

###
## Calling a sort function
###

mergeSort(arr)

###
## Output result
###

print(arr)

###
##  End of exemple of usage
###
